### PR TITLE
Add detailed error message infrastructure

### PR DESCRIPTION
**Pull request by BenjaminPelletier**
This PR begins to address #379 by:

1. Replacing all of our explicit Internal errors with `stacktrace` messages and propagations
1. Adding an error ID so a particular client-facing error message can be connected to backend logs quickly and easily, even if error message details are obfuscated
1. Fixing a number of issues with input validation and implicit Subscriptions exposed by the work above

In future PRs, we can add more `stacktrace.Propagate` statements to improve visibility into error causes, but that will be a large task and this PR aims to be a targeted introduction to the work.

### CODE REVIEW

File: pkg/errors/errors.go [hunk 1]
```diff
@@ -35,27 +36,53 @@ func init() {
...
-			logger.Error("encountered internal Status error during unary server call",
-				zap.String("method", info.FullMethod),
-				zap.Stringer("code", statusErr.Code()),
-				zap.String("message", statusErr.Message()),
-				zap.Any("details", statusErr.Details()),
-				zap.Error(err))
-			if statusErr.Code() == codes.Internal && obfuscateInternalErrors {
-				err = errInternal
+
+		if err != nil {
```
Comment (vosst) [inline pkg/errors/errors.go:54]: Please return early, helps with readability.

File: pkg/errors/errors.go [hunk 2]
```diff
@@ -35,27 +36,53 @@ func init() {
...
+			// Separate the root cause from the stacktrace wrapping.
+			trace := err.Error()
+			rootErr := stacktrace.RootCause(err)
+
+			statusErr, ok := status.FromError(rootErr)
+			switch {
+			case !ok:
+				errID := MakeErrID()
+				logger.Error(
+					fmt.Sprintf("encountered non-Status error %s during unary server call", errID),
```
Comment (vosst) [inline pkg/errors/errors.go:64]: `logger.Errorf` should give you the same functionality.

File: pkg/errors/errors.go [hunk 3]
```diff
@@ -35,27 +36,53 @@ func init() {
...
+					zap.String("stacktrace", trace),
+					zap.Error(rootErr))
+				if obfuscateInternalErrors {
+					err = status.Error(codes.Internal, fmt.Sprintf("Internal server error %s", errID))
+				} else {
+					err = status.Error(codes.Internal, err.Error())
+				}
+			case statusErr.Code() == codes.Internal, statusErr.Code() == codes.Unknown:
+				errID := MakeErrID()
+				logger.Error(
```
Comment (vosst) [inline pkg/errors/errors.go:75]: `logger.Errorf`.

File: pkg/errors/errors.go [hunk 4]
```diff
@@ -35,27 +36,53 @@ func init() {
 	}
 }
 
+func MakeErrID() string {
```
Comment (vosst) [inline pkg/errors/errors.go:39]: Please rename to `makeErrID()`. No need to export the function.

Comment (BenjaminPelletier) [reply pkg/errors/errors.go:39]: Done.

File: pkg/errors/errors.go (hunk 1 above)
Comment (BenjaminPelletier) [reply pkg/errors/errors.go:54]: Done.

File: pkg/errors/errors.go (hunk 2 above)
Comment (BenjaminPelletier) [reply pkg/errors/errors.go:64]: `type *zap.Logger has no field or method Errorf`; what did you have in mind?

Comment (BenjaminPelletier) [reply pkg/errors/errors.go]: More specifically: https://godoc.org/go.uber.org/zap#Logger

Comment (vosst) [reply pkg/errors/errors.go]: I was thinking about `zap.SugaredLogger` but it turns out that it does not support additional, unformatted fields. With that, comment resolved.

**Pull request merged**